agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -429,11 +429,157 @@ impl DeriveSchema for Stage {
+        ) -> LinkedHashMap<String, ProjectPathNode> {
+            let mut project_paths = ProjectPathNode {
+                children_nodes: LinkedHashMap::new(),
+                project_item: None,
+            };
+            for (k, v) in items
+                .iter()
+                .filter(|(_, p)| !matches!(p, ProjectItem::Exclusion))
+            {
+                // if the top level key for a path is not present in the root node, create an empty
+                // node for it
+                let path = k.split('.').map(|s| s.to_string()).collect::<Vec<String>>();
+                if !project_paths.children_nodes.contains_key(&path[0]) {
+                    project_paths.children_nodes.insert(
+                        path[0].clone(),
+                        ProjectPathNode {
+                            children_nodes: LinkedHashMap::new(),
+                            project_item: None,
+                        },
+                    );
+                }
+                // beginning at the root node, descend down the path, adding new nodes
+                // for any unencountered paths
+                let mut cur = &mut project_paths;
+                for (index, field) in path.iter().enumerate() {
+                    if !cur.children_nodes.contains_key(field) {
+                        cur.children_nodes.insert(
+                            field.clone(),
+                            ProjectPathNode {
+                                children_nodes: LinkedHashMap::new(),
+                                project_item: None,
+                            },
+                        );
+                    };
+                    cur = cur.children_nodes.get_mut(field).unwrap();
+                    // once we reach the terminal field for this path, set that node's
+                    // project item; i.e., for {$project: {"a.b.c": 1}}, once we reach "c",
+                    // that node's project_item should be Some(ProjectItem::Inclusion)
+                    if index == path.len() - 1 {
+                        cur.project_item = Some(v.clone());
+                    }
+                }
+            }
+            project_paths.children_nodes
+        }
+
+        /// get_schema_for_project_node takes a node and generates the schema for all
+        /// fields in the project by traversing top down and evaluating any project_items.
+        /// the parent schema keeps a reference to the schema for the current node from the
+        /// result set schema -- that is, we traverse the result set schema in parallel in
+        /// order to properly evaluate inclusions.
+        fn get_schema_for_project_node(
+            node: &ProjectPathNode,
+            parent_schema: Option<&Schema>,
+            state: &mut ResultSetState,
+        ) -> Schema {
+            if let Some(schema) = parent_schema {
+                // the former case is not likely, but if we have reached a terminal arm of the
```

Comment:
It's "not likely" or it's not possible? Can you give an example of how we could end up in this state?

Reply:
Oh wait, would is just be an empty `$project` stage?

Reply:
Yeah that is about correct. I just removed it, and added a new error for empty project, which will deserialize but we shouldn't entertain